[NETBEANS-1605] Removed NetBeans max HeapSize heuristics.

lkishalmi opened the pull request.

emilianbold commented:
For reference, the JVM default is documented here: https://docs.oracle.com/javase/8/docs/technotes/guides/vm/gctuning/parallel.html#default_heap_size although it's still not clear enough for me...

This commit is an improvement versus the status quo, but how many machines with under 4GB of RAM use NetBeans? And for those over 4GB we ask the same 1GB like the JVM. It's a draw.

I think we should actually ask for more RAM compared to the default as we were too conservative.

Considering 8GB is not so rare nowadays on a machine, I think asking 50% of the memory with a cap at 4GB sounds much better.

If we apply this commit we lose a way to increase the max heap.

Thus, I vote not for removing our heuristics, but for asking for more compared to the JVM default.

The other problem is that we have no nice way of allowing the user to increase the max heap size. Somehow editing a deep .conf file doesn't seem very user friendly.

matthiasblaesing commented:
The documentation does not seem to be correct. From my machine:

```
matthias@athena:~$ java -version
java version "1.8.0_181"
Java(TM) SE Runtime Environment (build 1.8.0_181-b13)
Java HotSpot(TM) 64-Bit Server VM (build 25.181-b13, mixed mode)
matthias@athena:~$ java -XshowSettings 2>&1 | grep Heap
    Max. Heap Size (Estimated): 3.46G
matthias@athena:~$ 
matthias@athena:~$ bin/jdk-11/bin/java -version
openjdk version "11" 2018-09-25
OpenJDK Runtime Environment 18.9 (build 11+28)
OpenJDK 64-Bit Server VM 18.9 (build 11+28, mixed mode)
matthias@athena:~$ bin/jdk-11/bin/java -XshowSettings 2>&1 | grep Heap
    Max. Heap Size (Estimated): 3.89G
matthias@athena:~$ 
```

That is also verified by using JMX on a jrunscript session. The java.lang/Memory/HeapMemoryUsage shows a maximum HeapMemoryUsage of 4173332480.

Given, that the permanent generation was moved offheap, 1/4 of available ram sound like a reasonable size.

lkishalmi commented:
Well, I have a died thread on this on the mailing list: http://mail-archives.apache.org/mod_mbox/netbeans-dev/201810.mbox/%3C8fb7cf19-c2d2-8147-a5c4-f5dcc84e0e80%40gmail.com%3E

I would favor the 25% as the JVM default, as it is often the case (at least for me) that you are running application servers databases an other IDE a resource hungry browser besides for NetBeans.
Anyway an option to select the JVM and the Max Heap Size shall be editable in the UI.

emilianbold commented:
Indeed, the documentation is confusing. So, Java max heap is by default 25% of RAM on systems with over 4GB or RAM, which sounds OK to me.

Since the default sounds better now, +1 on this PR.

emilianbold commented on nb/ide.launcher/windows/nblauncher.h:
```diff
@@ -64,7 +64,6 @@ class NbLauncher {
     virtual bool initBaseNames();
     virtual void addSpecificOptions(CmdArgs &args);
     virtual bool areWeOn32bits();
```
`areWeOn32bits` seems to exist just for `adjustHeapAndPermGenSize` so you should delete that one too.

lkishalmi commented:
Also removed the reference of CMS GC from netbeans.conf as it got deprecated in Java 9.
Removed the minimum heap size as of 32 megs, which would be the default when using Netbeans on a machine with 1Gb ram as it is the 1/64th of the available ram, it is often greater on modern machines. So removing that theoretically could speed up the boot performance.

lkishalmi commented:
@emilianbold please review.
